File: Sources/Olaf/UI/OlafUI.swift
import Foundation
import SwiftUI

/// General-purpose facade for OlafUI: shake → viewer setup, external tool registration, presentation.
///
/// ```swift
/// Olaf.start(.default)
/// OlafUI.install()       // viewer (shake) setup
/// ```
///
/// If you want to add your own custom diagnostic tool, use `install(tools:)` or `register(_:)`.
public enum OlafUI {

    /// Installs the observer that opens the viewer when the device is shaken, and registers the
    /// given external tools. Idempotent; call once.
    /// - Parameter tools: Custom external tool bridges the host wants to add.
    @MainActor
    public static func install(tools: [any ExternalToolBridge] = []) {
        #if canImport(UIKit)
        OlafPresenter.shared.installShakeObserver()
        #endif
        for tool in tools {
            ExternalToolRegistry.shared.register(tool)
        }
    }

    /// Registers a single custom external tool bridge. Becomes a switch-to button in the viewer.
    public static func register(_ bridge: any ExternalToolBridge) {
        ExternalToolRegistry.shared.register(bridge)
    }

    /// Removes all registered external tools.
    public static func unregisterAllTools() {
        ExternalToolRegistry.shared.removeAll()
    }

    /// Registers a handler for taps on the Olaf logo in the viewer's navigation bar.
    ///
    /// When a handler is set, the logo becomes a button: tapping it closes the viewer and invokes
    /// the handler **after** the viewer has fully closed, so it is safe to present another
    /// diagnostics tool from the handler. Useful when Olaf is installed alongside another
    /// shake-activated tool: shake opens Olaf first, and the logo hands off to the other tool.
    /// Pass `nil` to remove the handler (the logo becomes a plain image again).
    @MainActor
    public static func onLogoTap(_ handler: (() -> Void)?) {
        #if canImport(UIKit)
        OlafPresenter.shared.logoTapHandler = handler
        #endif
    }

    /// Opens the viewer programmatically.
    @MainActor
    public static func present() {
        #if canImport(UIKit)
        OlafPresenter.shared.present()
        #endif
    }

    /// Closes the viewer programmatically.
    /// - Parameter completion: Runs AFTER the viewer has fully closed. Self-presenting UIKit
    ///   tools should call their `show()` here (presenting before the dismiss animation finishes fails).
    @MainActor
    public static func dismiss(completion: (() -> Void)? = nil) {
        #if canImport(UIKit)
        OlafPresenter.shared.dismiss(completion: completion)
        #else
        completion?()
        #endif
    }

    /// Presents an embeddable SwiftUI tool modally over Olaf's own window.
    /// Returns to the Olaf viewer when dismissed.
    ///
    /// Self-presenting UIKit tools should use `dismiss()` + their own `show()` instead.
    @MainActor
    public static func presentExternal<Content: View>(@ViewBuilder _ content: () -> Content) {
        #if canImport(UIKit)
        OlafPresenter.shared.presentExternal(rootView: content())
        #endif
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/OlafUI.kt
package com.olaf.ui

import android.app.Application
import android.content.Context
import com.olaf.ui.bridge.ExternalToolBridge
import com.olaf.ui.bridge.ExternalToolRegistry
import com.olaf.ui.presentation.OlafPresenter

/**
 * Facade for the viewer: shake-to-open setup, external tool registration and programmatic
 * presentation.
 *
 * ```kotlin
 * Olaf.start(context)
 * OlafUI.install(application)   // shake → viewer
 * ```
 */
object OlafUI {

    /**
     * Installs the shake observer that opens the viewer, and registers the given external tools.
     * Idempotent; call once, typically from `Application.onCreate`.
     */
    fun install(application: Application, tools: List<ExternalToolBridge> = emptyList()) {
        OlafPresenter.install(application)
        tools.forEach(ExternalToolRegistry::register)
    }

    /** Registers a single external tool bridge; it becomes a button in the viewer. */
    fun register(bridge: ExternalToolBridge) {
        ExternalToolRegistry.register(bridge)
    }

    /** Removes every registered external tool. */
    fun unregisterAllTools() {
        ExternalToolRegistry.removeAll()
    }

    /**
     * Registers a handler for taps on the Olaf title in the viewer's app bar.
     *
     * With a handler set, the title becomes a button: tapping it closes the viewer and invokes the
     * handler, which makes it safe to present another diagnostics tool from there. Useful when
     * Olaf is installed alongside a second shake-activated tool — the shake opens Olaf, and the
     * title hands off. Pass `null` to remove the handler.
     */
    fun onLogoTap(handler: (() -> Unit)?) {
        OlafPresenter.logoTapHandler = handler
    }

    /** Opens the viewer programmatically. */
    fun present(context: Context? = null) {
        OlafPresenter.present(context)
    }

    /** Closes the viewer programmatically. */
    fun dismiss() {
        OlafPresenter.dismiss()
    }
}
